src/main/java/com/thealgorithms/maths/MatrixRank.java

```diff
@@ -0,0 +1,154 @@
+package com.thealgorithms.maths;
+
+/**
+ * This class provides a method to compute the rank of a matrix.
+ * In linear algebra, the rank of a matrix is the maximum number of linearly independent rows or columns in the matrix.
+ * For example, consider the following 3x3 matrix:
+ * 1 2 3
+ * 2 4 6
+ * 3 6 9
+ * Despite having 3 rows and 3 columns, this matrix only has a rank of 1 because all rows (and columns) are multiples of each other.
+ * It's a fundamental concept that gives key insights into the structure of the matrix.
+ * It's important to note that the rank is not only defined for square matrices but for any m x n matrix.
+ *
+ * @author Anup Omkar
+ */
+public final class MatrixRank {
+
+    private final static double EPSILON = 1e-10;
+
+    /**
+     * @brief Computes the rank of the input matrix
+     *
+     * @param matrix The input matrix
+     * @return The rank of the input matrix
+     */
+    public static int computeRank(double[][] matrix) {
+        validateInputMatrix(matrix);
+
+        int numRows = matrix.length;
+        int numColumns = matrix[0].length;
+        int rank = 0;
+
+        boolean[] rowMarked = new boolean[numRows];
+
+        double[][] matrixCopy = new double[numRows][numColumns];
+        for (int rowIndex = 0; rowIndex < numRows; ++rowIndex) {
+            System.arraycopy(matrix[rowIndex], 0, matrixCopy[rowIndex], 0, numColumns);
+        }
```

Comment:
Please move it into a separate method, like `deepCopy`.

Reply:
Changed it.

Reply:
@anupomkar could you please add this test checking that the input is not modified? In general, it is a good practice go _express requirements as tests_.

Reply:
Is that really needed?
Firstly, all the helper functions like isZero, isJaggred, checkInput and deepCopy should not be part of this algorithm/class. They all should be moved to some common place or helper class where all these helper functionalities must reside and tested. The `MatrixRankTest.java` class should not have responsibility of testing these things. It should only test main business logic / algorithm logic.

Reply:
In my opinion: it is really needed.

Your `MatrixRank.computeRank` seems to be a reasonable _unit_ and it is reasonable to require that it will:
- return correct result,
- throw an error for wrong input,
- not modify the input.

So these aspects should be tested.

Should the mentioned methods be part of this class or different? Depends: if there are other places in this repo that exactly the same functions are needed, then yes - but this if out of scope for this pull-request. I do not see a relation between their location and the tests of your class. Even if the method `isJaggred` would be defined in another class and testes separately, you would still need to check if `MatrixRank.computeRank` throws for an input not being a proper matrix.

Reply:
Added unit tests. Please check